gslib/tests/test_cp.py

```diff
@@ -1700,6 +1700,31 @@ def _CopyAndCheck():
 
     _CopyAndCheck()
 
+  def test_copy_bucket_to_bucket_with_duplicate_nested_object_names(self):
+    """Tests copying from bucket to bucket preserves file structure."""
+    src_bucket_uri = self.CreateBucket()
+    dst_bucket_uri = self.CreateBucket()
+    self.CreateObject(bucket_uri=src_bucket_uri,
+                      object_name='dir1/file.txt',
+                      contents=b'data')
+    self.CreateObject(bucket_uri=src_bucket_uri,
+                      object_name='dir2/file.txt',
+                      contents=b'data')
+
+    # Use @Retry as hedge against bucket listing eventual consistency.
+    @Retry(AssertionError, tries=3, timeout_secs=1)
+    def _CopyAndCheck():
+      self.RunGsUtil(
+          ['cp', '-R',
+           suri(src_bucket_uri) + '/*',
+           suri(dst_bucket_uri)])
```

Comment:
I think a similar test should be created for non-existing destination dir as well, because that's the key change that this PR is introducing.

Reply:
bucket will get treated as an existing dir.